util/auth.py

```diff
@@ -6,9 +6,11 @@
 from aplus_auth.payload import Permission
 from django.core.exceptions import PermissionDenied
 from django.http.response import HttpResponse
+from django.urls import reverse_lazy
 
 ViewType = Callable[..., HttpResponse]
-login_required: Callable[[ViewType],ViewType] = login_required_base(redirect_url="/login?referer={url}")
+login_required: Callable[[ViewType], ViewType] = login_required_base(
+    redirect_url=reverse_lazy('login', kwargs={'referer': '{url}'}))
```

Comment:
Does this really work for you? I get a crash on the MOOC-Grader front page when I haven't previously logged into the grader:

```
NoReverseMatch at /

Reverse for 'login' with keyword arguments '{'referer': '{url}'}' not found. 1 pattern(s) tried: ['login\\Z']
```

The URL query parameters (after the `?` question mark) are not part of the Django URL resolver. The parameters need to be appended to the output of the `reverse()` function. Since `reverse_lazy` does not return a normal string due to the lazy evaluation, then maybe you need to create a new lazy string that appends the URL query parameters to the output after `reverse_lazy()`'s lazy string has been evaluated.

Read these too:
* https://stackoverflow.com/questions/4995279/including-a-querystring-in-a-django-core-urlresolvers-reverse-call
* https://stackoverflow.com/questions/29252312/how-can-i-join-lazy-translation-in-django

Does this work? The StackOverflow link above suggests using `format_lazy()` for concatenating lazy strings.
```python
login_required: Callable[[ViewType], ViewType] = login_required_base(
    redirect_url=format_lazy('{}?{}', reverse_lazy('login'), 'referer={url}'),
)
```
https://docs.djangoproject.com/en/3.2/ref/utils/#django.utils.text.format_lazy

I'm not sure if the `login_required_base()` function works correctly with the lazy parameter `redirect_url`. Try it out!
https://github.com/apluslms/aplus-auth/blob/8050cecd99dd9335adf684a2ec57c068e7a67edd/src/aplus_auth/auth/django.py#L41

Reply:
Login required_base seems to work with the lazy formatting